packages/near-lake-primitives/src/types/block.ts

```diff
@@ -0,0 +1,159 @@
+import { Action, isActionReceipt, outcomeWithReceiptToReceipt, receiptViewToAction } from './receipts';
+import { StreamerMessage, ValidatorStakeView } from './core/types';
+import { Receipt } from './receipts';
+import { Transaction } from './transactions';
+import { Event, logToRawEvent } from './events';
+import { fromStateChangeViewToStateChange, StateChange } from './stateChanges';
+
+export class Block {
+    readonly streamerMessage: StreamerMessage;
+    private executedReceipts: Receipt[];
+    readonly postponedReceipts: Receipt[];
+    readonly transactions: Transaction[];
+    private _actions: Map<string, Action>;
+    private _events: Map<string, Event[]>;
+    private _stateChanges: StateChange[];
+
+    constructor(streamerMessage: StreamerMessage, executedReceipts: Receipt[], postponedReceipts: Receipt[],
+        transactions: Transaction[], actions: Map<string, Action>,
+        events: Map<string, Event[]>, stateChanges: StateChange[]) {
+        this.streamerMessage = streamerMessage;
+        this.executedReceipts = executedReceipts;
+        this.postponedReceipts = postponedReceipts;
+        this.transactions = transactions;
+        this._actions = actions;
+        this._events = events;
+        this._stateChanges = stateChanges;
+    }
+
+    get blockHash(): string {
+        return this.header().hash;
+    }
+
+    get prevBlockHash(): string {
+        return this.header().prevHash;
+    }
+    get blockHeight(): number {
+        return this.header().height;
+    }
+
+    header(): BlockHeader {
+        return streamerMessageToBlockHeader(this.streamerMessage);
+    }
+
+    receipts(): Receipt[] {
+        if (this.executedReceipts.length == 0) {
+            this.executedReceipts = this.streamerMessage.shards.flatMap((shard) => shard.receiptExecutionOutcomes).map((executionReceipt) => outcomeWithReceiptToReceipt(executionReceipt))
+        }
+        return this.executedReceipts;
+    }
+
+    actions(): Action[] {
+        const actions: Action[] = this.streamerMessage.shards.flatMap((shard) => shard.receiptExecutionOutcomes).filter((exeuctionOutcomeWithReceipt) => isActionReceipt(exeuctionOutcomeWithReceipt.receipt)).map((exeuctionOutcomeWithReceipt) => receiptViewToAction(exeuctionOutcomeWithReceipt.receipt)).filter((action) => action !== null).map(action => action as Action)
+        return actions
+    }
+
+    events(): Event[] {
+        const events = this.receipts().flatMap((executedReceipt) => executedReceipt.logs.map(logToRawEvent).map((rawEvent) => {
+            if (rawEvent) {
+                let event = { relatedReceiptId: executedReceipt.receiptId, rawEvent: rawEvent } as Event
+                return event
+            } else {
+                return null
+            }
+        })).filter((event) => event !== null).map((event) => event as Event)
```

Comment:
Same as above, use a type predicate instead of casting.

Reply:
Agree with you here. 

```
.filter((event: Event | null) => event !== null).map((event:  Event) => event)
```

We can talk this one over a brief call. I've tried to specify the type for event in the map, but typescript does not detect the fact that since I filtered out the null values, it is just a Event now.